dotAsset is Breaking FileViewStrategy

### Parent Issue

_No response_

### User Story

In our content API v1 we are using the new Transformers which is great.  The issue is that our API responses for binaries are all different based on what type of content you have and what field type you are using.  This is part of what makes using images in dotCMS - all the responses vary.

This does not need to be the case as in the end everything is just a binary field.


### The Bug
The bug is this - if you create 2 related content types, parent + child and the child has a File or Image field that has a dotAsset in it, and you request the parent content via the v1 version of the content api with `?depth=1`, it will throw this error:

```
{
  "message": "Contentlet : 3fb168f8-34ce-4f9c-875f-4a2a9b13c9c3 is not a FileAsset"
}
```

Likewise, In the block editor, if you add a link a piece of content that has a relationship to another piece of content that has a file field that has a dotAsset selected, you will see this error and your content will not render or be displayed. 




 

### Acceptance Criteria

Fix the bug and make sure our API responses for dotAssets, FileAssets, Binaries and file and image fields are the same

### Proposed Objective

Technical User Experience

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

```
Caused by: com.dotmarketing.business.DotStateException: Contentlet : 3fb168f8-34ce-4f9c-875f-4a2a9b13c9c3 is not a FileAsset
	at com.dotmarketing.portlets.fileassets.business.FileAssetAPIImpl.fromContentlet(FileAssetAPIImpl.java:256) ~[classes/:?]
	at com.dotmarketing.portlets.contentlet.transform.strategy.FileViewStrategy.transform(FileViewStrategy.java:106) ~[classes/:?]
	at com.dotmarketing.portlets.contentlet.transform.strategy.FileViewStrategy.transform(FileViewStrategy.java:79) ~[classes/:?]
	at com.dotmarketing.portlets.contentlet.transform.strategy.AbstractTransformStrategy.apply(AbstractTransformStrategy.java:91) ~[classes/:?]
	at com.dotmarketing.portlets.contentlet.transform.DotContentletTransformerImpl.lambda$transform$3(DotContentletTransformerImpl.java:105) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:720) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:558) ~[?:?]
	at com.dotmarketing.portlets.contentlet.transform.DotContentletTransformerImpl.transform(DotContentletTransformerImpl.java:107) ~[classes/:?]
	at com.dotmarketing.portlets.contentlet.transform.DotContentletTransformerImpl.lambda$hydrate$7(DotContentletTransformerImpl.java:121) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at com.dotmarketing.portlets.contentlet.transform.DotContentletTransformerImpl.hydrate(DotContentletTransformerImpl.java:123) ~[classes/:?]
	at com.dotcms.rest.ContentResource.contentletToJSON(ContentResource.java:1408) ~[classes/:?]
	at com.dotcms.rest.ContentResource.addRelatedContentToJsonArray(ContentResource.java:1296) ~[classes/:?]
	at com.dotcms.rest.ContentResource.addRelationshipsToJSON(ContentResource.java:1211) ~[classes/:?]
	at com.dotcms.rendering.velocity.viewtools.content.util.ContentUtils.addRelationships(ContentUtils.java:814) ~[classes/:?]
```

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/transform/strategy/FileViewStrategy.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/transform/strategy/FileViewStrategy.java
@@ -90,7 +90,8 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
         final String fileAssetIdentifier = (String) contentlet.get(field.variable());
 
         if (!UtilMethods.isSet(fileAssetIdentifier)) {
-            return null;
+            Logger.warn(this, "FileAsset identifier is empty for field: " + field.variable());
+            return Map.of();
         }
 
         Optional<Contentlet> fileAsContentOptional = APILocator.getContentletAPI()
@@ -99,17 +100,20 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
                         , contentlet.getLanguageId(),
                         APILocator.systemUser(), true);
 
-        FileAsset fileAsset = null;
+        if(fileAsContentOptional.isEmpty()) {
+            //Prevent NPE
+            Logger.warn(this, "Live FileAsset not found for identifier: " + fileAssetIdentifier);
+            return Map.of();
+        }
 
-        if(fileAsContentOptional.isPresent()) {
-            final FileAssetAPI fileAssetAPI = APILocator.getFileAssetAPI();
-            //This does always assume we're getting a fileAsset we don't want to miss a dotAsset
-            final Contentlet fileLikeContentlet = fileAsContentOptional.get();
-            if(fileLikeContentlet.isDotAsset()){
-                fileAsset = fileAsset(fileLikeContentlet, fileAssetAPI);
-            } else {
-                fileAsset = fileAssetAPI.fromContentlet(fileLikeContentlet);
-            }
+        final FileAsset fileAsset;
+        final FileAssetAPI fileAssetAPI = APILocator.getFileAssetAPI();
+        //This does always assume we're getting a fileAsset we don't want to miss a dotAsset
+        final Contentlet incoming = fileAsContentOptional.get();
+        if(incoming.isDotAsset()){
+            fileAsset = convertToFileAsset(incoming, fileAssetAPI);
+        } else {
+            fileAsset = fileAssetAPI.fromContentlet(incoming);
         }
 
         final Map<String, Object> map = new HashMap<>();
@@ -158,7 +162,7 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
             map.put(FILE_NAME_FIELD, fileName);
             map.put("fileSize", fileSize);
         }
-        map.put("type", "file_asset");
+        map.put("type", assetType(incoming));
         map.put("isContentlet", true);
         return map;
     }
@@ -170,7 +174,7 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
      * @return the fileAsset
      * @throws DotDataException if the content type is not found
      */
-    private  FileAsset fileAsset(final Contentlet dotAsset, final FileAssetAPI api) throws DotDataException {
+    private  FileAsset convertToFileAsset(final Contentlet dotAsset, final FileAssetAPI api) throws DotDataException {
         final ContentType contentType =
                Try.of(()->APILocator.getContentTypeAPI(APILocator.systemUser()).find("FileAsset")).getOrNull();
 
@@ -179,7 +183,18 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
         }
 
         final Contentlet newCon = new Contentlet(dotAsset);
+        //here we're simply replacing the content type with the fileAsset content type to bypass a check that takes place in the fromContentlet method
+        //No big deal since we're not going to save this contentlet
         newCon.setContentType(contentType);
         return api.fromContentlet(newCon);
+    }
+
+    /**
+     * This method will return the asset type
+     * @param contentlet the contentlet
+     * @return the asset type
+     */
+    String assetType(final Contentlet contentlet){
+        return contentlet.isDotAsset() ? "dot_asset" : "file_asset";
     }
 }

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/transform/strategy/FileViewStrategy.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/transform/strategy/FileViewStrategy.java
@@ -1,7 +1,7 @@
 package com.dotmarketing.portlets.contentlet.transform.strategy;
 
-import static com.dotmarketing.portlets.contentlet.transform.strategy.TransformOptions.LOAD_META;
 import static com.dotmarketing.portlets.contentlet.transform.strategy.TransformOptions.AVOID_MAP_SUFFIX_FOR_VIEWS;
+import static com.dotmarketing.portlets.contentlet.transform.strategy.TransformOptions.LOAD_META;
 import static com.dotmarketing.portlets.contentlet.transform.strategy.TransformOptions.USE_ALIAS;
 import static com.dotmarketing.portlets.fileassets.business.FileAssetAPI.DESCRIPTION;
 import static com.dotmarketing.portlets.fileassets.business.FileAssetAPI.FILE_NAME_FIELD;
@@ -12,7 +12,6 @@ import static com.dotmarketing.portlets.fileassets.business.FileAssetAPI.UNDERLY
 import static com.dotmarketing.util.UtilHTML.getIconClass;
 import static com.dotmarketing.util.UtilHTML.getStatusIcons;
 import static com.dotmarketing.util.UtilMethods.getFileExtension;
-import static com.dotmarketing.util.UtilMethods.isImage;
 import static com.dotmarketing.util.UtilMethods.isNotSet;
 import static com.dotmarketing.util.UtilMethods.isSet;
 import static com.liferay.util.StringPool.BLANK;
@@ -21,14 +20,14 @@ import com.dotcms.api.APIProvider;
 import com.dotcms.contenttype.model.field.Field;
 import com.dotcms.contenttype.model.field.FileField;
 import com.dotcms.contenttype.model.field.ImageField;
+import com.dotcms.contenttype.model.type.ContentType;
 import com.dotmarketing.beans.Identifier;
 import com.dotmarketing.business.APILocator;
-import com.dotmarketing.business.CacheLocator;
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotSecurityException;
-import com.dotmarketing.portlets.contentlet.business.ContentletCache;
 import com.dotmarketing.portlets.contentlet.model.Contentlet;
 import com.dotmarketing.portlets.fileassets.business.FileAsset;
+import com.dotmarketing.portlets.fileassets.business.FileAssetAPI;
 import com.dotmarketing.util.Logger;
 import com.dotmarketing.util.UtilMethods;
 import com.liferay.portal.model.User;
@@ -103,7 +102,14 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
         FileAsset fileAsset = null;
 
         if(fileAsContentOptional.isPresent()) {
-            fileAsset = APILocator.getFileAssetAPI().fromContentlet(fileAsContentOptional.get());
+            final FileAssetAPI fileAssetAPI = APILocator.getFileAssetAPI();
+            //This does always assume we're getting a fileAsset we don't want to miss a dotAsset
+            final Contentlet fileLikeContentlet = fileAsContentOptional.get();
+            if(fileLikeContentlet.isDotAsset()){
+                fileAsset = fileAsset(fileLikeContentlet, fileAssetAPI);
+            } else {
+                fileAsset = fileAssetAPI.fromContentlet(fileLikeContentlet);
+            }
         }
 
         final Map<String, Object> map = new HashMap<>();
@@ -155,5 +161,25 @@ public class FileViewStrategy extends AbstractTransformStrategy<Contentlet> {
         map.put("type", "file_asset");
         map.put("isContentlet", true);
         return map;
+    }
+
+    /**
+     * This method will convert a dotAsset to a fileAsset
+     * @param dotAsset the dotAsset to be converted
+     * @param api the fileAssetAPI
+     * @return the fileAsset
+     * @throws DotDataException if the content type is not found
+     */
+    private  FileAsset fileAsset(final Contentlet dotAsset, final FileAssetAPI api) throws DotDataException {
+        final ContentType contentType =
+               Try.of(()->APILocator.getContentTypeAPI(APILocator.systemUser()).find("FileAsset")).getOrNull();
+
+        if(null == contentType){
+            throw new DotDataException("FileAsset content type not found");
+        }
+
+        final Contentlet newCon = new Contentlet(dotAsset);
+        newCon.setContentType(contentType);
+        return api.fromContentlet(newCon);
     }
 }

--- a/dotcms-integration/src/test/java/com/dotmarketing/portlets/contentlet/transform/ContentletTransformerTest.java
+++ b/dotcms-integration/src/test/java/com/dotmarketing/portlets/contentlet/transform/ContentletTransformerTest.java
@@ -2,16 +2,20 @@ package com.dotmarketing.portlets.contentlet.transform;
 
 import com.dotcms.api.APIProvider;
 import com.dotcms.api.APIProvider.Builder;
+import com.dotcms.contenttype.model.field.BinaryField;
 import com.dotcms.contenttype.model.field.ConstantField;
 import com.dotcms.contenttype.model.field.Field;
 import com.dotcms.contenttype.model.field.FieldBuilder;
+import com.dotcms.contenttype.model.field.FileField;
 import com.dotcms.contenttype.model.field.ImageField;
 import com.dotcms.contenttype.model.field.StoryBlockField;
 import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.contenttype.model.type.ContentType;
 import com.dotcms.contenttype.model.type.FileAssetContentType;
 import com.dotcms.datagen.CategoryDataGen;
+import com.dotcms.datagen.ContentTypeDataGen;
 import com.dotcms.datagen.ContentletDataGen;
+import com.dotcms.datagen.FieldDataGen;
 import com.dotcms.datagen.SiteDataGen;
 import com.dotcms.datagen.TestDataUtils;
 import com.dotcms.datagen.TestDataUtils.TestFile;
@@ -398,6 +402,49 @@ public class ContentletTransformerTest extends BaseWorkflowIntegrationTest {
         Assert.assertNotEquals(map2.get("mimeType"),"unknown");
         Assert.assertTrue((long)map2.get("size") > 0);
 
+    }
+
+    /**
+     * Given Scenario: We have a contentlet with a dotAsset field, and we push it into the transformer prepared with the FILEASSET_VIEW strategy.
+     * Expected Result: We should get the dotAsset field with the expected properties.
+     * @throws Exception
+     */
+    @Test
+    public void Test_DotAsset_Field_Pushed_Into_Transformer() throws Exception {
+
+        final Contentlet dotAssetLikeContentlet = TestDataUtils.getDotAssetLikeContentlet();
+        ContentletDataGen.publish(dotAssetLikeContentlet);
+
+        final List<Field> fields = List.of(
+                new FieldDataGen()
+                        .name("title")
+                        .velocityVarName("title")
+                        .next(),
+                new FieldDataGen()
+                        .type(FileField.class)
+                        .name("dotAsset")
+                        .velocityVarName("dotAsset")
+                        .next()
+        );
+
+        final String contentTypeName =  "withDotAssetType"+System.currentTimeMillis();
+
+        final ContentType contentType = new ContentTypeDataGen()
+                .name(contentTypeName)
+                .velocityVarName(contentTypeName)
+                .fields(fields)
+                .nextPersisted();
+
+        final Contentlet contentlet = new ContentletDataGen(contentType.id())
+                .setProperty("title", "Test")
+                .setProperty("dotAsset", dotAssetLikeContentlet.getIdentifier())
+                .nextPersistedAndPublish();
+
+        final Contentlet transformed = new DotTransformerBuilder().hydratedContentMapTransformer().content(contentlet).build().hydrate().get(0);
+        Map<?,?> asset = (Map)transformed.getMap().get("dotAsset");
+        Assert.assertNotNull(asset);
+        Assert.assertFalse(asset.isEmpty());
+        Assert.assertEquals("dot_asset", asset.get("type"));
     }
 
 

